runtime/src/leader_bank_status.rs

```diff
@@ -0,0 +1,167 @@
+use {
+    crate::bank::Bank,
+    solana_sdk::slot_history::Slot,
+    std::{
+        sync::{Arc, Condvar, Mutex, RwLock, Weak},
+        time::{Duration, Instant},
+    },
+};
+
+/// Tracks leader status of the validator node and notifies when:
+///     1. A leader slot begins
+///     2. A leader slot completes
+#[derive(Debug, Default)]
+pub struct LeaderBankStatus {
+    /// Current state of the system
+    status: Mutex<Status>,
+    /// Weak reference to the current bank
+    bank: RwLock<Option<(Slot, Weak<Bank>)>>,
+    /// CondVar to notify status changes and waiting
+    condvar: Condvar,
+}
+
+/// Leader status state machine for the validator:
+/// [Unininitialized] -> [InProgress] -> [Completed] --|
+///                          ^-------------------------|
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+enum Status {
+    /// Initial state, no bank, but also not completed yet
+    #[default]
+    Uninitialized,
+    /// Slot is in progress as leader
+    InProgress,
+    /// PoH has reached the end of the slot, and the next bank as leader is not available yet
+    Completed,
+}
+
+impl LeaderBankStatus {
+    /// Set the status to `InProgress` and notify any waiting threads
+    /// if the status was not already `InProgress`.
+    pub fn set_in_progress(&self, bank: &Arc<Bank>) {
+        let mut status = self.status.lock().unwrap();
+        if matches!(*status, Status::InProgress) {
+            return;
+        }
+
+        *status = Status::InProgress;
+        *self.bank.write().unwrap() = Some((bank.slot(), Arc::downgrade(bank)));
+        drop(status);
+
+        self.condvar.notify_all();
+    }
+
+    /// Set the status to `Completed` and notify any waiting threads
+    /// if the status was not already `Completed`
+    /// and the slot is higher than the current slot (sanity check).
+    pub fn set_completed(&self, slot: Slot) {
+        let mut status = self.status.lock().unwrap();
+        if matches!(*status, Status::Completed) {
+            return;
+        }
+
+        if let Some((current_slot, _)) = *self.bank.read().unwrap() {
+            if slot < current_slot {
+                return;
+            }
+        }
+
+        *status = Status::Completed;
+        drop(status);
+
+        self.condvar.notify_all();
+    }
+
+    /// If the status is `InProgress`, immediately return a weak reference to the bank.
+    /// Otherwise, wait up to the `timeout` for the status to become `InProgress`.
+    /// Returns `None` if the timeout is reached.
+    pub fn wait_for_in_progress(&self, timeout: Duration) -> Option<Weak<Bank>> {
```

Comment:
`Option`-ing `Weak` looks a bit odd. I think `Weak::new()` can be used to flatten this return type? https://doc.rust-lang.org/std/sync/struct.Weak.html#method.new

Reply:
> ..., so maybe we can just use `new` as you suggested since there's no reason to distinguish. What do you think?

yeah, let's do that.

Reply:
done: 15858f01955267825ad7392648b53c6f85d01d95